test/macros.jl

```diff
@@ -80,6 +80,47 @@ end
     @test constraint_object(c).set == MOI.EqualTo(0.0)
 end
 
+struct NewVariable <: JuMP.AbstractVariable
+    info::JuMP.VariableInfo
+end
+
+@testset "Extension variables constrained on creation #2594" begin
+    function JuMP.build_variable(
+        _error::Function,
+        info::JuMP.VariableInfo,
+        ::Type{NewVariable},
+    )
+        return NewVariable(info)
+    end
+    function JuMP.add_variable(model::Model, v::NewVariable, name::String = "")
+        return JuMP.add_variable(model, ScalarVariable(v.info), name)
+    end
+    function JuMP.add_variable(
+        model::Model,
+        v::VariablesConstrainedOnCreation{
+            MOI.SecondOrderCone,
+            VectorShape,
+            NewVariable,
+        },
+        names,
+    )
+        vs = map(i -> ScalarVariable(i.info), v.scalar_variables)
+        new_v = VariablesConstrainedOnCreation(vs, v.set, v.shape)
+        return JuMP.add_variable(model, new_v, names)
+    end
+
+    model = Model()
+    @variable(model, 0 <= x <= 1, NewVariable, Bin)
+    @test lower_bound(x) == 0
+    @test upper_bound(x) == 1
+    @test is_binary(x)
+    @test name(x) == "x"
+
+    @variable(model, y[1:3] in SecondOrderCone(), NewVariable)
+    @test name.(y) == ["y[$i]" for i in 1:3]
+    @test num_constraints(model, Vector{VariableRef}, MOI.SecondOrderCone) == 1
```

Comment:
This test doesn't test that `NewVariable` actually gets used.

If you did something like:
```Julia
function JuMP.add_variable(model::Model, v::NewVariable, name::String = "")
    return JuMP.add_variable(model, ScalarVariable(v.info), "NewVariable" * name)
end
```
then you could check the code-path actually got hit by looking at the variable names.

Reply:
I have updated the tests accordingly.